### PR TITLE
Yadda 20150107 rc

### PR DESCRIPTION
**Pull request by ryancowan**
Old PR from yadda part 1: https://github.com/jmmastey/level_up_exercises/pull/577

This pull request is for yadda parts 2 and 3. It also includes the feedback from yadda part 1.

Files for part 2:
yadda/yadda-views.sql

These views were initially written with a subquery to fetch the related ratings information and performance on them was terrible when using `explain analyze`. I received feedback from my team after a code review and ended up removing the subqueries in favor of a regular join on the ratings table. I didn't do this initially because I was getting tripped up by having to include additional IDs in the `group by` clause.

The explain on these view queries shows that the scanning and sorting are costly. The sequential scan that happens on the ratings table is also expensive.

I really like this tool for helping to visualize explains: http://explain.depesz.com/

### CODE REVIEW

File: yadda/data-generator.rb [hunk 1]
```diff
@@ -0,0 +1,101 @@
+require 'pg'
+require 'faker'
```
Comment (paulh16) [inline yadda/data-generator.rb]: Good use of faker.

File: yadda/data-generator.rb [hunk 2]
```diff
@@ -0,0 +1,101 @@
...
+
+def main
+  generate_person_records
+  generate_brewery_records
+  generate_style_records
+  generate_beer_records
+  generate_rating_records
+end
+
+def generate_rating_records count=1000
```
Comment (paulh16) [inline yadda/data-generator.rb]: I'd be consistent throughout the code about putting parentheses around method parameters.

File: yadda/data-generator.rb [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+    }
+  end
+end
+
+def get_record_id table, primary_key
+  record = @conn.exec( "SELECT #{primary_key} FROM #{table} ORDER BY RANDOM() LIMIT 1" ).first
+  record["#{primary_key}"]
+end
+
+def insert_record table, info_hash
```
Comment (paulh16) [inline yadda/data-generator.rb]: I really like this method you extracted!

File: yadda/data-generator.rb [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+      birthday: Faker::Date.between("1900-01-01", Date.today),
+      description: Faker::Lorem.sentence(3),
+      created_at: Faker::Date.between("2014-01-01", Date.today),
+      updated_by: 1
+    }
+  end
+end
+
+def get_record_id table, primary_key
+  record = @conn.exec( "SELECT #{primary_key} FROM #{table} ORDER BY RANDOM() LIMIT 1" ).first
```
Comment (paulh16) [inline yadda/data-generator.rb]: Agree with Hound on the space needed inside the parentheses.

File: yadda/yadda-indices.sql [hunk 5]
```diff
@@ -0,0 +1,13 @@
+-- Usage: psql -d yadda -f yadda/yadda-indices.sql
+
+-- indices
+CREATE UNIQUE INDEX breweries_id_index ON breweries(brewery_id);
+CREATE INDEX beers_id_to_breweries_id_index ON beers(beer_id, brewery_id);
+CREATE INDEX persons_id_index ON persons(person_id);
+CREATE UNIQUE INDEX ratings_id_index ON rating(rating_id);
```
Comment (paulh16) [inline yadda/yadda-indices.sql]: This is a good index.

File: yadda/yadda-indices.sql [hunk 6]
```diff
@@ -0,0 +1,13 @@
+-- Usage: psql -d yadda -f yadda/yadda-indices.sql
+
+-- indices
+CREATE UNIQUE INDEX breweries_id_index ON breweries(brewery_id);
+CREATE INDEX beers_id_to_breweries_id_index ON beers(beer_id, brewery_id);
```
Comment (paulh16) [inline yadda/yadda-indices.sql]: Good index.

File: yadda/yadda-indices.sql [hunk 7]
```diff
@@ -0,0 +1,13 @@
...
+CREATE INDEX persons_id_index ON persons(person_id);
+CREATE UNIQUE INDEX ratings_id_index ON rating(rating_id);
+
+-- multicolumn indices
+CREATE INDEX persons_to_beers_index ON ratings(person_id, beer_id);
+
+-- attempt at functional index
+-- CREATE INDEX rating_avg_computed_idx ON rating (AVG(overall));
```
Comment (paulh16) [inline yadda/yadda-indices.sql]: Are you trying to create an index for the average of the ratings? If so, something along the lines of the following would help:

`CREATE INDEX overall_ratings_index ON rating (((look + smell + taste + feel) / 4));`

File: yadda/yadda-views.sql [hunk 8]
```diff
@@ -0,0 +1,37 @@
+-- Usage: psql -d yadda -f yadda/yadda-views.sql
+
+-- Top beers from a given brewery, according to their total ratings. 
+create or replace view top_beers AS
```
Comment (paulh16) [inline yadda/yadda-views.sql]: Great view and nice use of joins over subqueries.

https://robots.thoughtbot.com/postgresql-performance-considerations

File: yadda/yadda-views.sql [hunk 9]
```diff
@@ -0,0 +1,37 @@
...
+  left join ratings on ratings.beer_id = beers.beer_id
+  group by breweries.brewery_id, beers.beer_id
+  order by ratings_count desc;
+
+-- "Recent score" for a beer, where only ratings within the last six months are counted and the ratings within that period are averaged. 
+create or replace view recent_score AS
+  select breweries.brewery_id as breweryid,
+    breweries.name as breweryname,
+    beers.name as beername,
+    AVG(ratings.overall) as computed
```
Comment (paulh16) [inline yadda/yadda-views.sql]: I'd be interested as to the performance of this and whether you could use a functional index that precomputes this average computation.

File: yadda/yadda-views.sql [hunk 10]
```diff
@@ -0,0 +1,37 @@
...
+    beers.name as beername,
+    COUNT(ratings.overall) as ratings_count
+  from breweries 
+  left join beers on beers.brewery_id = breweries.brewery_id
+  left join ratings on ratings.beer_id = beers.beer_id
+  group by breweries.brewery_id, beers.beer_id
+  order by ratings_count desc;
+
+-- "Recent score" for a beer, where only ratings within the last six months are counted and the ratings within that period are averaged. 
+create or replace view recent_score AS
```
Comment (paulh16) [inline yadda/yadda-views.sql]: Overall, good job on this view.

File: yadda/yadda-views.sql [hunk 11]
```diff
@@ -0,0 +1,37 @@
...
+
+-- "You might also enjoy", which picks beers of the same style with high average scores, and then sorts them randomly. 
+create or replace view also_enjoy AS
+  select beers.name as beername,
+    styles.style as beerstyle,
+    AVG(ratings.overall) as computed
+  from beers
+  left join styles on beers.style_id = styles.style_id
+  left join ratings on ratings.beer_id = beers.beer_id
+  group by styles.style_id, beers.beer_id
```
Comment (paulh16) [inline yadda/yadda-views.sql]: Consider adding a `HAVING` clause that will only return records with a overall rating above a certain number.

**Comment by paulh16** [comment]
@ashtonthomas @djkotowski @ryancowan 

Overall, this is excellent work. I really like the use of joins over subqueries for performance reasons. 

Part 2 is complete, and for Parts 3 and 4, that I’d like to see the EXPLAINs of the queries with your analysis of the results, and then to add indices to address performance concerns. You have already added some indices, so you are on your way to finishing this up. 

See the article below; this will also help you on Parts 3 and 4:

https://robots.thoughtbot.com/postgresql-performance-considerations

Nice work! :+1:

File: yadda/yadda-views.sql (hunk 8 above)
Comment (djkotowski) [inline yadda/yadda-views.sql]: Also it's a very small thing, but be consistent with the capitalization of your SQL commands.

File: yadda/yadda-views.sql (hunk 9 above)
Comment (djkotowski) [inline yadda/yadda-views.sql]: I'd pick a more specific name for this than `computed`.

**Comment by djkotowski** [comment]
This looks good.  You're ready for parts 3 and 4.

**Pull request closed without merging**